core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1733,16 +1733,31 @@ class KafkaController(val config: KafkaConfig,
       else brokerEpoch
 
     val shouldSkipShutdownSafetyCheck = controllerContext.skipShutdownSafetyCheck.getOrElse(id, -1L) >= actualBrokerEpoch
+    var passedAllShutdown: Boolean = true // For logging purpose
+    if (controllerContext.getLivePreferredControllerIds.size <= config.liMinPreferredControllerCount
+        // Only check if this broker affects live preferred controller set so that for non-preferred-controller clusters it doesn't get blocked
```

Comment:
Didn't quite get how this check can rule out non-prefere-controller clusters. Could you elaborate a bit?

Reply:
Do you mean for non-prefered-controller clusters, the liMinPreferredControllerCount is set to a very small value like 0?

Reply:
>  liMinPreferredControllerCount is set to a very small value like 0?
That's a more of an orthogonal angle.

Only preferred controller would be in `controllerContext.getLivePreferredControllerIds`.  So if I'm shutting down either
1. data broker
2. a controller in a cluster that doesn't configure preferred controller
3. a "fallback" controller (i.e. preferred controllers are down, but this cluster allows data broker to take controller for availability reason)

Those would not fall in the to if bock.